packages/server/lib/cloud/protocol.ts

```diff
@@ -1,71 +1,101 @@
+import { createGzip } from 'zlib'
+import fetch from 'cross-fetch'
 import { performance } from 'perf_hooks'
 
+const routes = require('./routes')
+const pkg = require('@packages/root')
+const { agent } = require('@packages/network')
 const debug = Debug('cypress:server:protocol')
 const debugVerbose = Debug('cypress-verbose:server:protocol')
 
-const setupProtocol = async (url?: string): Promise<AppCaptureProtocolInterface | undefined> => {
-  let script: string | undefined
+export class ProtocolManager implements ProtocolManagerShape {
+  private _runId?: string
+  private _instanceId?: string
+  private _db?: Database.Database
+  private _dbPath?: string
+  private _errors: ProtocolError[] = []
+  private _protocol: AppCaptureProtocolInterface | undefined
 
-  // TODO(protocol): We will need to remove this option in production
-  if (process.env.CYPRESS_LOCAL_PROTOCOL_PATH) {
-    script = await fs.readFile(process.env.CYPRESS_LOCAL_PROTOCOL_PATH, 'utf8')
-  } else if (url) {
-    // TODO(protocol): Download the protocol script from the cloud
+  get protocolEnabled (): boolean {
+    return !!this._protocol
   }
 
-  if (script) {
-    const cypressProtocolDirectory = path.join(os.tmpdir(), 'cypress', 'protocol')
+  async setupProtocol (script: string, runId: string) {
+    debug('setting up protocol via script')
+    try {
+      this._runId = runId
+      if (script) {
+        const cypressProtocolDirectory = path.join(os.tmpdir(), 'cypress', 'protocol')
 
-    // TODO(protocol): Handle any errors here appropriately. Likely, we will want to handle all errors in the initialization process similarly (e.g. downloading, file permissions, etc.)
-    await fs.ensureDir(cypressProtocolDirectory)
-    const vm = new NodeVM({
-      console: 'inherit',
-      sandbox: {
-        Debug,
-        performance: {
-          now: performance.now,
-          timeOrigin: performance.timeOrigin,
-        },
-      },
-    })
+        await fs.ensureDir(cypressProtocolDirectory)
+        const vm = new NodeVM({
+          console: 'inherit',
+          sandbox: {
+            Debug,
+            performance: {
+              now: performance.now,
+              timeOrigin: performance.timeOrigin,
+            },
+          },
+        })
 
-    const { AppCaptureProtocol } = vm.run(script)
+        const { AppCaptureProtocol } = vm.run(script)
 
-    return new AppCaptureProtocol()
+        this._protocol = new AppCaptureProtocol()
+      }
+    } catch (error) {
+      this._errors.push({
+        error,
+        args: [script],
+        captureMethod: 'setupProtocol',
+      })
+    }
   }
 
-  return
-}
-
-class ProtocolManagerImpl implements ProtocolManager {
-  private protocol: AppCaptureProtocolInterface | undefined
-
-  async setupProtocol (url?: string) {
-    debug('setting up protocol via url %s', url)
-
-    this.protocol = await setupProtocol(url)
-  }
+  async connectToBrowser (cdpClient: CDPClient) {
+    const newCdpClient: CDPClient = {
```

Comment:
```suggestion
    // Wrap the cdp client listeners so that we can be notified of any errors that may occur
    const newCdpClient: CDPClient = {
```